### ISSUE TITLE
"Connection refused" when proxying UDP port

### ISSUE THREAD

**Issue by justadri** [mholt/caddy-l4#182]
Hello,

I am trying to proxy a UDP port using caddy-l4. However, when I try to connect to the service thru Caddy, I a "read: connection refused" error in my Caddy logs. I have confirmed that the port is listening on the machine hosting the service ("athena"):
```
sudo ss -tulpn | grep 3478
udp   UNCONN 0      0            0.0.0.0:3478       0.0.0.0:*    users:(("docker-proxy",pid=300380,fd=4))                 
udp   UNCONN 0      0               [::]:3478          [::]:*    users:(("docker-proxy",pid=300386,fd=4)) 
```

and I can connect to it from the host Caddy is running on ("terminus"):
```
nc -vuzw 3 athena.dgtlu.net 3478
Connection to athena.domain.com (192.168.0.21) 3478 port [udp/*] succeeded!
```

The exact error I am seeing:
```
{
  "level": "error",
  "ts": 1712207073.016409,
  "logger": "layer4.handlers.proxy",
  "msg": "upstream connection",
  "local_address": "192.168.0.30:37876",
  "remote_address": "192.168.0.20:3478",
  "error": "read udp 192.168.0.30:37876->192.168.0.20:3478: read: connection refused"
}
```

Any idea what I am doing wrong?

My caddyfile.json:
```
{
	"logging": {
		"logs": {
			"default": {
				"writer": {
					"filename": "/var/log/caddy/default.log",
					"output": "file"
				},
				"level": "INFO",
				"exclude": [
					"http.log.access.log0",
					"http.log.access.log1",
					"http.log.access.log2",
					"http.log.access.log3",
					"http.log.access.log4",
					"http.log.access.log5",
					"http.log.access.log6",
					"http.log.access.log7",
					"http.log.access.log8",
					"http.log.access.log9"
				]
			},
			"log0": {
				"writer": {
					"filename": "/var/log/caddy/agent_access.log",
					"output": "file"
				},
				"level": "DEBUG",
				"include": [
					"http.log.access.log0"
				]
			},
			"log1": {
				"writer": {
					"filename": "/var/log/caddy/filerun_access.log",
					"output": "file"
				},
				"level": "DEBUG",
				"include": [
					"http.log.access.log1"
				]
			},
			"log2": {
				"writer": {
					"filename": "/var/log/caddy/office_access.log",
					"output": "file"
				},
				"level": "DEBUG",
				"include": [
					"http.log.access.log2"
				]
			},
			"log3": {
				"writer": {
					"filename": "/var/log/caddy/librephoto_access.log",
					"output": "file"
				},
				"level": "DEBUG",
				"include": [
					"http.log.access.log3"
				]
			},
			"log4": {
				"writer": {
					"output": "stderr"
				},
				"level": "DEBUG",
				"include": [
					"http.log.access.log4"
				]
			},
			"log5": {
				"writer": {
					"output": "stderr"
				},
				"level": "DEBUG",
				"include": [
					"http.log.access.log5"
				]
			},
			"log6": {
				"writer": {
					"filename": "/var/log/caddy/stream_access.log",
					"output": "file"
				},
				"level": "DEBUG",
				"include": [
					"http.log.access.log6"
				]
			},
			"log7": {
				"writer": {
					"filename": "/var/log/caddy/trillium_access.log",
					"output": "file"
				},
				"level": "DEBUG",
				"include": [
					"http.log.access.log7"
				]
			},
			"log8": {
				"writer": {
					"filename": "/var/log/caddy/agendav_access.log",
					"output": "file"
				},
				"level": "DEBUG",
				"include": [
					"http.log.access.log8"
				]
			},
			"log9": {
				"writer": {
					"filename": "/var/log/caddy/baikal_access.log",
					"output": "file"
				},
				"level": "DEBUG",
				"include": [
					"http.log.access.log9"
				]
			}
		}
	},
	"apps": {
		"crowdsec": {
			"api_url": "http://localhost:8080/",
			"api_key": "M/jkKKkUyGfZBwsuGOnaCPUxkawo+XslmpIJYp+URyk",
			"ticker_interval": "15s",
			"enable_streaming": true,
			"enable_hard_fails": false
		},
		"layer4": {
			"servers": {
				"srv1": {
					"listen": ["udp/:3478"],
					"routes": [
						{
							"handle": [
								{
									"handler": "proxy",
									"upstreams": [
										{"dial": ["udp/persephone.domain.com:3478"]}
									]
								}
							]
						}
					]
				}
			}
		},
		"http": {
			"servers": {
				"srv0": {
					"listen": [
						":443"
					],
					"routes": [
						{
							"match": [
								{
									"host": [
										"office.domain.com"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "reverse_proxy",
													"upstreams": [
														{
															"dial": "persephone.domain.com:8081"
														}
													]
												}
											]
										}
									]
								}
							],
							"terminal": true
						},
						{
							"match": [
								{
									"host": [
										"photos.domain.com"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "reverse_proxy",
													"upstreams": [
														{
															"dial": "persephone.domain.com:2283"
														}
													]
												}
											]
										}
									]
								}
							],
							"terminal": true
						},
						{
							"match": [
								{
									"host": [
										"stream.domain.com"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "reverse_proxy",
													"upstreams": [
														{
															"dial": "persephone.domain.com:8180"
														}
													]
												}
											]
										}
									]
								}
							],
							"terminal": true
						},
						{
							"match": [
								{
									"host": [
										"agent.domain.com"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "reverse_proxy",
													"headers": {
														"request": {
															"set": {
																"Host": [
																	"{http.request.host}"
																]
															}
														}
													},
													"upstreams": [
														{
															"dial": "athena.domain.com:8090"
														}
													]
												}
											]
										}
									]
								}
							],
							"terminal": true
						},
						{
							"match": [
								{
									"host": [
										"files.domain.com"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "vars",
													"root": "/var/www/html"
												}
											]
										},
										{
											"handle": [
												{
													"handler": "headers",
													"response": {
														"set": {
															"Strict-Transport-Security": [
																"max-age=15552000;"
															]
														}
													}
												}
											],
											"match": [
												{
													"path": [
														"/*"
													]
												}
											]
										},
										{
											"handle": [
												{
													"handler": "reverse_proxy",
													"upstreams": [
														{
															"dial": "persephone.domain.com:8080"
														}
													]
												},
												{
													"browse": {},
													"handler": "file_server",
													"hide": [
														"./Caddyfile"
													]
												}
											]
										}
									]
								}
							],
							"terminal": true
						},
						{
							"match": [
								{
									"host": [
										"notes.domain.com"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "reverse_proxy",
													"upstreams": [
														{
															"dial": "persephone.domain.com:8888"
														}
													]
												}
											]
										}
									]
								}
							],
							"terminal": true
						},
						{
							"match": [
								{
									"host": [
										"plex.domain.com"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "headers",
													"response": {
														"set": {
															"Referrer-Policy": [
																"no-referrer-when-downgrade"
															],
															"Strict-Transport-Security": [
																"max-age=31536000;"
															],
															"X-Content-Type-Options": [
																"nosniff"
															],
															"X-Frame-Options": [
																"DENY"
															],
															"X-Xss-Protection": [
																"1"
															]
														}
													}
												},
												{
													"encodings": {
														"gzip": {},
														"zstd": {}
													},
													"handler": "encode",
													"prefer": [
														"gzip",
														"zstd"
													]
												},
												{
													"handler": "reverse_proxy",
													"upstreams": [
														{
															"dial": "persephone.domain.com:32400"
														}
													]
												}
											]
										}
									]
								}
							],
							"terminal": true
						},
						{
							"match": [
								{
									"host": [
										"hass.domain.com"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "reverse_proxy",
													"upstreams": [
														{
															"dial": "iris.domain.com:8123"
														}
													]
												}
											]
										}
									]
								}
							],
							"terminal": true
						},
						{
							"match": [
								{
									"host": [
										"sync.domain.com"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"match": [
												{
													"path": [
														"/calexa*"
													]
												}
											],
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "reverse_proxy",
													"rewrite": {
														"strip_path_prefix": "/calexa"
													},
													"upstreams": [
														{
															"dial": "persephone.domain.com:5000"
														}
													]
												}
											]
										},
										{
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "reverse_proxy",
													"upstreams": [
														{
															"dial": "persephone.domain.com:8008"
														}
													]
												}
											]
										}
									]
								}
							],
							"terminal": true
						},
						{
							"match": [
								{
									"host": [
										"cal.domain.com"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"handle": [
												{
													"handler": "crowdsec"
												},
												{
													"handler": "reverse_proxy",
													"upstreams": [
														{
															"dial": "persephone.domain.com:8088"
														}
													]
												}
											]
										}
									]
								}
							],
							"terminal": true
						}
					],
					"logs": {
						"logger_names": {
							"agent.domain.com": "log0",
							"cal.domain.com": "log8",
							"files.domain.com": "log1",
							"hass.domain.com": "log5",
							"notes.domain.com": "log7",
							"office.domain.com": "log2",
							"photos.domain.com": "log3",
							"plex.domain.com": "log4",
							"stream.domain.com": "log6",
							"sync.domain.com": "log9"
						}
					}
				}
			}
		},
		"tls": {
			"automation": {
				"policies": [
					{
						"subjects": [
							"office.domain.com",
							"photos.domain.com",
							"stream.domain.com",
							"agent.domain.com",
							"files.domain.com",
							"notes.domain.com",
							"plex.domain.com",
							"hass.domain.com",
							"sync.domain.com",
							"cal.domain.com"
						],
						"issuers": [
							{
								"ca": "https://acme.zerossl.com/v2/DV90",
								"email": "domains@domain.com",
								"module": "zerossl"
							}
						]
					}
				]
			}
		}
	}
}

```

**Comment by mohammed90** [comment]
>  I can connect to it from the host Caddy is running on (\"terminus\"):\n\nnc -vuzw 3 athena.dgtlu.net 3478\nConnection to athena.domain.com (192.168.0.21) 3478 port [udp/*] succeeded!\nThe exact error I am seeing:\n\n{\n  \"level\": \"error\",\n  \"ts\": 1712207073.016409,\n  \"logger\": \"layer4.handlers.proxy\",\n  \"msg\": \"upstream connection\",\n  \"local_address\": \"192.168.0.30:37876\",\n  \"remote_address\": \"192.168.0.20:3478\",\n  \"error\": \"read udp 192.168.0.30:37876->192.168.0.20:3478: read: connection refused\"\n}\nAny idea what I am doing wrong?

The upstream address you're using in Caddy is `persephone.domain.com:3478`, which apparently resolves to `192.168.0.20`.s. However, the address you're using in nc is `athena.dgtlu.net` and resolves to `192.168.0.21`.

**Comment by justadri** [comment]
@mohammed90 oh my gosh i can't believe i missed that! thank you so much for the extra pair of eyes!
